### DISCUSSION TITLE
Unix Domain Socket

### DISCUSSION THREAD

**Discussion by hemalbavishi** [eclipse-iceoryx/iceoryx#1404] [Q&A] [answered]
Hello,

I am bit analyzing Iceoryx implementation for IPC Communication based on UDS. Following are my observation:
1. Application creates two UDS Socket - One for Send and One for Receive 
2. Roudi creates n+1 UDS Socket - One for Receive and n for Send

I have following questions based on above understanding. Request you to clarify the same. 

1. Since UDS is full duplex (unlike Pipe which is half duplex), isn't possible to send/receive data using same socket? 
2. Assume that I have multiple threads in the User Application(Process) which is trying to subscribe and offer different services almost at the same time. In that case, each thread will send the data to Roudi almost at the same time over UDS and wait for Roudi to respond or acknowledge. Is there any chance of data loss if multiple requests are sent from the User Application via Iceoryx Runtime Interface to Roudi and response sent to Iceoryx Run time interface from Roudi almost at the same time?

**Comment by elfenpiff** [comment]
@hemalbavishi 

> Since UDS is full duplex (unlike Pipe which is half duplex), isn't possible to send/receive data using same socket?

Yes it is in theory possible. But there is a problem when you send out a message and want to receive an answer to that message on the same unix domain socket. How long do you have to wait for the corresponding process to read the message? If you wait not long enough it is possible that you read your own message again when you call `recvfrom` for instance. If you wait too long you have a massive latency in your application.

To your observation:
* RouDi creates one unix domain socket (let's call it `A`) where every application can send requests, like acquiring ports or other resources
* Every application creates a unix domain socket (let's call it `app`) named like the runtime name you provide during runtime initialization.
* Every application opens socket `A` on startup to acquire resources and the answer to the resource request is sent by roudi on the `app`

> Assume that I have multiple threads in the User Application(Process) which is trying to subscribe and offer different services almost at the same time. In that case, each thread will send the data to Roudi almost at the same time over UDS and wait for Roudi to respond or acknowledge. Is there any chance of data loss if multiple requests are sent from the User Application via Iceoryx Runtime Interface to Roudi and response sent to Iceoryx Run time interface from Roudi almost at the same time?

This is a very good question. At the moment I would refrain from that use case but a quick look at the source code gives me a good feeling that this should be thread-safe but to be honest, we did not yet consider this use case. I will create an issue for that to investigate this further, write maybe some test so that we really ensure that this valid use case is working.

**Reply by elBoberido** [reply]
@elfenpiff I think you have misunderstood the question because we once used a message queue which is not bi-directional. With the unix domain socket it would not be a problem but then we would need to restrict ourselves to a bi-directional channel. For some platforms we might either not have this or rather use a non bi-directional channel.

@hemalbavishi I need to double check, but if I remember correctly, the unix domain socket would block on the application side if RouDi is too slow with the messages and no message would be lost.

**Reply by hemalbavishi** [reply]
@elBoberido Now I understand why Iceoryx has used different sockets for bi-directional communication as some platforms e.g Windows might not have similar feature available.

**Comment by elfenpiff** [comment, marked as answer]
@hemalbavishi 

I created the issue #1405

**Reply by hemalbavishi** [reply]
Thank you! Closing this issue as question is answered.
